python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -149,6 +159,91 @@ def _read_file(
+    def _set_encoding_ratio(self, encoding_ratio: int):
+        """Set image file encoding ratio, to provide accurate size in bytes metadata."""
+        self._encoding_ratio = encoding_ratio
+
+    def _get_block_metadata(
+        self,
+        paths: List[str],
+        schema: Optional[Union[type, "pyarrow.lib.Schema"]],
+        *,
+        rows_per_file: Optional[int],
+        file_sizes: List[Optional[int]],
+    ) -> BlockMetadata:
+        metadata = super()._get_block_metadata(
+            paths, schema, rows_per_file=rows_per_file, file_sizes=file_sizes
+        )
+        if metadata.size_bytes is not None:
+            metadata.size_bytes = int(metadata.size_bytes * self._encoding_ratio)
+        return metadata
+
+
+class _ImageFolderDatasourceReader(_FileBasedDatasourceReader):
+    def __init__(
+        self,
+        delegate: FileBasedDatasource,
+        paths: List[str],
+        filesystem: "pyarrow.fs.FileSystem",
+        partition_filter: PathPartitionFilter,
+        meta_provider: _ImageFileMetadataProvider = _ImageFileMetadataProvider(),
+        **reader_args,
+    ):
+        super().__init__(
+            delegate=delegate,
+            paths=paths,
+            filesystem=filesystem,
+            schema=None,
+            open_stream_args=None,
+            meta_provider=meta_provider,
+            partition_filter=partition_filter,
+            **reader_args,
+        )
+        self._encoding_ratio = self._estimate_files_encoding_ratio()
+        meta_provider._set_encoding_ratio(self._encoding_ratio)
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        return sum(self._file_sizes) * self._encoding_ratio
+
+    def _estimate_files_encoding_ratio(self) -> float:
+        """Return an estimate of the image files encoding ratio."""
+        # Prefetch one image file into memory, and use it to estimate data size for
+        # all images, because all images are homogeneous with same size after resizing.
+        import pandas as pd
+
+        start_time = time.perf_counter()
+
+        # Filter out empty file to avoid noise.
+        non_empty_path_and_size = list(
+            filter(lambda p: p[1] > 0, zip(self._paths, self._file_sizes))
+        )
+        assert len(non_empty_path_and_size) > 0
+        # Prefetch first non-empty image file to do estimation.
+        path = non_empty_path_and_size[0][0]
+        single_image_block: pd.DataFrame = self._delegate._read_file(
+            self._filesystem.open_input_stream(path),
+            path,
+            self._reader_args.get("root"),
+            self._reader_args.get("size"),
+        )
+        single_image_memory_size = single_image_block.memory_usage().sum()
+        total_estimated_memory_size = single_image_memory_size * len(
+            non_empty_path_and_size
+        )
+        total_file_size = sum(p[1] for p in non_empty_path_and_size)
+        ratio = total_estimated_memory_size / total_file_size
```

Comment:
> Maybe add a TODO to sample multiple images? Compression ratio could vary quite a lot depending on image content. But this seems more like a P1.

@ericl - all images are same size after resizing, and `single_image_memory_size` is the in-memory data size for one image after resizing. I assume each image's in-memory data size should be very similar given resizing, right? The ratio is calculate based on total file sizes. Do you think it still needs sampling multiple images?

Reply:
Well, it depends on the image. For example, JPG images could be compressed with varying compression quality, which can vary by 1-2 orders of magnitude: https://www.graphicsmill.com/blog/2014/11/06/Compression-ratio-for-different-JPEG-quality-values

Reply:
Ah, got it, this makes sense. I was thinking they were still encoded in memory for some reason.

Reply:
Yep I think the single file sample should be fine.

You technically could have that one file is dictionary-encoded in memory while another is not, but that discrepancy shouldn't arise through a single read call in Datasets: when reading Parquet, Arrow will only automatically choose dictionary encoding for partition columns; the user will have to opt-in for all others, in which case the column will be dictionary-encoded for all files.

Arrow only does dynamic dictionary encoding (starting with dictionary encoding and falling back to plain encoding if the dictionary grows too large) when writing Parquet to disk, not when reading Parquet into memory.

Reply:
@clarkzinzow - this is image data source, not Parquet, and we are [storing image as NumPy array in memory](https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/image_folder_datasource.py#L130). So I assume it should not have any dictionary encoding, right? (at least I didn't find out based on testing several exampel images, but I may be missing anything)

Reply:
Ah @c21 yep my bad, I was switching between different PRs and I didn't realize this was for the image folder datasource! 😅

Reply:
@clarkzinzow - no worries, and I agree with what you said for Parquet! 😄